src/cautiousrobot/__main__.py

```diff
@@ -0,0 +1,271 @@
+# Downloads images from CSV with option to save downsampled copies.
+# Logs image downloads and failures in json files.
+# Logs response codes as strings, not int64.
+
+# Logs saved in same folder as CSV used for download
+# Downsized images are saved in <img_dir>_downsized
+
+import requests
+import shutil
+import io
+import pandas as pd
+import argparse
+import hashlib
+
+from tqdm import tqdm
+import os
+import sys
+import time
+from PIL import Image
+from sumbuddy import get_checksums
+from cautiousrobot.utils import log_response, update_log
+
+
+REDO_CODE_LIST = [429, 500, 502, 503, 504]
+
+
+def parse_args():
+    available_algorithms = ', '.join(hashlib.algorithms_available)
+
+    parser = argparse.ArgumentParser()
+    # Use argument groups for required vs optional (both get short flags too) https://bugs.python.org/issue9694#msg132327
+    # Required arguments
+    req_args = parser.add_argument_group("required arguments")
+    req_args.add_argument("-i", "--input-file", required = True, help = "path to CSV file with urls.", nargs = "?")
+    req_args.add_argument("-o", "--output-dir", required = True, help = "main directory to download images into.", nargs = "?")
+    
+    # Optional arguments
+    opt_args = parser.add_argument_group("optional arguments")
+    opt_args.add_argument("-s", "--subdir-col", required = False,
+                        help = "name of column to use for subfolders in image directory (defaults to flat directory if left blank)",
+                        nargs = "?")
+    opt_args.add_argument("-n", "--img-name-col", default = "filename", help = "column to use for image filename (default: filename)", nargs = "?")
+    opt_args.add_argument("-u", "--url-col", default = "file_url", help = "column with URLs to download (default: file_url)", nargs = "?")
+    opt_args.add_argument("-w", "--wait-time", default = 3, help = "seconds to wait between tries (default: 3)", type = int)
+    opt_args.add_argument("-r", "--max-retries", default = 5, help = "max times to retry download on a single image (default: 5)", type = int)
+    opt_args.add_argument("-l", "--side-length", required = False,
+                        help = "number of pixels per side for resized square images (default: no resized images created)",
+                        type = int)
+    opt_args.add_argument("-x", "--starting-idx", default = 0, help = "index of CSV at which to start download (default: 0)", type = int)
+    opt_args.add_argument("-a", "--checksum-algorithm", default = 'md5', #choices = available_algorithms,
+                        help = f"checksum algorithm to use on images (default: md5, available: {available_algorithms})"
+                        )
+
+    return parser.parse_args()
+
+
+def download_images(data, img_dir, log_filepath, error_log_filepath, filename = "filename",
```

Comment:
The code in this function is rather indented and a little long. Consider creating an issue to break this up to improve readability. The inline comments you already have might be a good way to think about breaking it up.

Reply:
Very true. I think implementing the retries with `urllib3` or `HTTPAdapter` (as mentioned [above](https://github.com/Imageomics/cautious-robot/pull/1#discussion_r1626727747)), might also help with that. Curious about your take on that question.

Reply:
Since you are already using requests I would go with the requests docs(`HTTPAdapter`):
https://requests.readthedocs.io/en/latest/user/advanced/#example-automatic-retries
It looks like it uses `urllib3` under the hood so it's pretty much the same thing.

One area where this approach might fail you is on a streaming response where you don't get the entire file. When that happens you get a successful http response, but the server doesn't return enough data. Typically there is a header with the file size to check for this. For that you would need a more complicated retry logic.  In the past I have used [tenacity](https://tenacity.readthedocs.io/en/latest/) and it was ok.

Reply:
That sounds ok. I would document memory requirements: they need enough memory to hold the largest image they download.

Reply:
That sounds doable; we'll be updating to streaming for a full version moving forward.